Support knn filter on nested metadata (#2315)
elastic/docs-content: solutions/search/vector/knn.md

}
```

##### Filtering on nested metadata [nested-knn-search-filtering-nested-metatadata]
```{applies_to}
stack: ga 9.2
```

Below we filter on nested metadata.
When scoring parents' documents this query only considers vectors that
have "paragraph.language" set to "EN".

```console
POST passage_vectors/_search
{
    "fields": [
        "full_text"
    ],
    "_source": false,
    "knn": {
        "query_vector": [0.45, 45],
        "field": "paragraph.vector",
        "k": 2,
        "filter": {
            "match": {
                "paragraph.language": "EN"
            }
        }
    }
}
```

> Below we have two filters: one over nested metadata
> and another over the top level metadata. For scoring parents' documents,
> this query only considers vectors that have "paragraph.language" set to "EN" 
> and whose parents have creation time within the request range.

Reviewer: ```suggestion
The following query has two filters: one over nested metadata and
another over the top-level metadata. When scoring parent documents, this
query only considers vectors that have "paragraph.language" set to "EN"
and whose parent documents were created within the specified range.
```

        "k": 2,
        "filter": [
            {"match": {"paragraph.language": "EN"}},
            {"range": { "creation_time": { "gte": "2019-05-01", "lte": "2019-05-05"}}}
        ]
    }
}
```

#### Filtering by sibling nested fields in nested KNN search [nested-knn-search-filtering-sibling]
```{applies_to}
stack: ga 9.2
```

Nested knn search also allows to do pre-filtering on sibling nested fields.
For example, given "paragraphs" and "metadata" as nested fields, we can search "paragraphs.vector" and filter by "metadata.key" and "metadata.value".

```console
POST passage_vectors/_search
{
    "fields": [
        "full_text"
    ],
    "_source": false,
    "knn": {
        "query_vector": [0.45, 45],
        "field": "paragraph.vector",
        "k": 2,
        "filter": {
            "nested": {
                "path": "metadata",
                "query": {
                    "bool": {
                        "must": [
                            { "match": { "metadata.key": "author" } },
                            { "match": { "metadata.value": "Doe" } }
                        ]
                    }
                }
            }
        }
    }
}
```

> Retrieving "inner_hits" when filtering on sibling nested fiels is not supported.

Reviewer: ```suggestion
:::{note}
Retrieving "inner_hits" when filtering on sibling nested fields is not supported.
:::
```

This sentence feels lonely all on its own 😄
